metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/tracing/ConsensusTracers.scala

```diff
@@ -0,0 +1,40 @@
+package io.iohk.metronome.hotstuff.service.tracing
+
+import cats.implicits._
+import io.iohk.metronome.tracer.Tracer
+import io.iohk.metronome.hotstuff.consensus.ViewNumber
+import io.iohk.metronome.hotstuff.consensus.basic.{
+  Agreement,
+  Event,
+  ProtocolError,
+  QuorumCertificate
+}
+
+case class ConsensusTracers[F[_], A <: Agreement](
+    timeout: Tracer[F, ViewNumber],
+    newView: Tracer[F, ViewNumber],
+    quorum: Tracer[F, QuorumCertificate[A]],
+    fromPast: Tracer[F, Event.MessageReceived[A]],
+    fromFuture: Tracer[F, Event.MessageReceived[A]],
+    stashed: Tracer[F, ProtocolError.TooEarly[A]],
+    rejected: Tracer[F, ProtocolError[A]],
+    error: Tracer[F, Throwable]
+)
+
+object ConsensusTracers {
+  import ConsensusEvent._
+
+  def apply[F[_], A <: Agreement](
+      tracer: Tracer[F, ConsensusEvent[A]]
+  ): ConsensusTracers[F, A] =
+    ConsensusTracers[F, A](
+      timeout = tracer.contramap[ViewNumber](Timeout(_)),
+      newView = tracer.contramap[ViewNumber](NewView(_)),
```

Comment:
The contramapping of tracers came from Cardano, as explained in [this seminar](https://drive.google.com/file/d/11j19P5QbAdHe4Yz9FAc7irEvPikMpnv8/view). I assume you are familar with that and asking about why there is this `ConsensusTracers` thing with a tracer per event type, rather than just passing the `Tracer[F, ConsensusEvent[A]]`. 

I had the following motivations:
1. A bit more friendly syntax, not having to instantiate the event type directly but just call a method that can do something before mapping to the event. For example we could trace a block ID, but retrieve the full block and project out a trace for all included transactions if we had to. A bit more massaging is happening in [network tracers](https://github.com/input-output-hk/metronome/blob/3fe4bf72afb3b8e3bce1eff52ebfc076e9811b58/metronome/networking/src/io/iohk/metronome/networking/NetworkTracers.scala), which would be ugly if it appeared everywhere.
2. Having every event separately allows finer modifications to how some of them are to be traced. For example if we want to also emit some metrics to a particular event, we can use `copy` and overwrite just some of the fields. 

To support multiple arguments I [opted](https://github.com/input-output-hk/metronome/blob/3fe4bf72afb3b8e3bce1eff52ebfc076e9811b58/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/tracing/ConsensusTracers.scala#L16) for a tuple, so every field can stay a `Tracer[F, <something>]` which means I we can still use the `Tracer` combinators like `|+|` to append others to it.

Reply:
Thanks, that answers my question